packages/contracts/src/contracts/current/forwarder/MixinBuyExactTokens.sol

```diff
@@ -0,0 +1,128 @@
+    {
+        address makerTokenAddress = readAddress(orders[0].makerAssetData, 1); 
+        // We can short cut here for effeciency and use buyFeeTokensInternal if maker asset token is ZRX
+        // this buys us exactly that amount taking into account the fees saving gas
+        Exchange.FillResults memory requestedTokensResult;
+        if (makerTokenAddress == address(ZRX_TOKEN)) {
+            requestedTokensResult = buyFeeTokensInternal(orders, signatures, tokenAmount);
+            require(requestedTokensResult.makerAssetFilledAmount >= tokenAmount, "traded amount did not meet acceptable threshold");
+        } else {
+            Exchange.FillResults memory expectedMarketBuyFillResults = expectedMaketBuyFillResults(orders, tokenAmount);
+            if (expectedMarketBuyFillResults.takerFeePaid > 0) {
+                // Fees are required for these orders. Buy enough ZRX to cover the future market buy
+                Exchange.FillResults memory feeTokensResult = buyFeeTokensInternal(feeOrders, feeSignatures, expectedMarketBuyFillResults.takerFeePaid);
+                totalFillResults.takerAssetFilledAmount = feeTokensResult.takerAssetFilledAmount;
+                totalFillResults.takerFeePaid = feeTokensResult.takerFeePaid;
+            }
+            // Make our market buy of the requested tokens with the remaining balance
+            requestedTokensResult = EXCHANGE.marketBuyOrders(orders, tokenAmount, signatures);
+            require(requestedTokensResult.makerAssetFilledAmount == tokenAmount, "traded amount did not meet acceptable threshold");
+        }
+        addFillResults(totalFillResults, requestedTokensResult);
+        // Transfer all tokens to msg.sender
+        transferToken(makerTokenAddress, msg.sender, totalFillResults.makerAssetFilledAmount);
+        return totalFillResults;
+    }
+
+    function buyNFTTokensInternal(
+        Order[] memory orders,
+        bytes[] memory signatures,
+        Order[] memory feeOrders,
+        bytes[] memory feeSignatures,
+        uint256 takerTokenAmount)
+        private
+        returns (Exchange.FillResults memory totalFillResults)
+    {
+        uint256 totalFeeAmount;
+        for (uint256 i = 0; i < orders.length; i++) {
+            totalFeeAmount = safeAdd(totalFeeAmount, orders[i].takerFee);
+        }
+        if (totalFeeAmount > 0) {
+            // Fees are required for these orders. Buy enough ZRX to cover the future fill
+            Exchange.FillResults memory feeTokensResult = buyFeeTokensInternal(feeOrders, feeSignatures, totalFeeAmount);
+            totalFillResults.takerFeePaid = feeTokensResult.takerFeePaid;
+            totalFillResults.takerAssetFilledAmount = feeTokensResult.takerAssetFilledAmount;
+        }
+        for (uint256 n = 0; n < orders.length; n++) {
+            // Fail if it wasn't fully filled otherwise we will keep WETH
+            Exchange.FillResults memory fillOrderResults = EXCHANGE.fillOrKillOrder(
```

Comment:
Can we use `batchFillOrKillOrders` here?

Reply:
We could but we would also have to for loop and transfer the tokens back. Unless spender and receiver were separated.